docs/proposals/podpolicy.md

```diff
@@ -0,0 +1,149 @@
+// an empty PodSelector (matches all pods in the namespace).
+type PodPolicySpec struct {
+	// A list of rules that define the policy. Each rule is tested in order, and the first one
+	// that matches is applied. 
+	Rules []PodPolicyRule `json:"rules,omitempty"`
+}
+
+// If the PolicyPredicate is met, then all of the actions specified
+// by the remaining fields are applied. 
+type PodPolicyRule struct {
+	PolicyPredicate *PolicyPredicate `json:"policyPredicate,omitempty"`
+	// Add the specified annotations to the pod.
+	Annotations map[string]string  `json:annotations,omitempty`
+	// In the future, we will add more actions here.
+}
+
+// QoS describes the Quality of Service class of a pod. Kubelet defines QoS on a
+// per-container, per-resource basis, but here we need to match a single QoS per pod.
+// We assume the consumer of PodPolicy will compute the per-pod QoS as the max QoS
+// across all resources and containers in the pod.
+type QoS string
+
+const (
+	QoSGuaranteed QoS = "Guaranteed"
+	QoSBurstable  QoS = "Burstable"
+	QoSBestEffort QoS = "BestEffort"
+)
+
+// All criteria must be met for the PolicyPredicate to be considered met.
+type PolicyPredicate struct {
+	// Note that Namespace is not listed here because PodPolicy is per-Namespace.
+	PodSelector *PodSelector `json:"podSelector,omitempty"`
+	QoS QoS  `json:qos,omitempty`
+}
+
+// PodPolicy describes the PodPolicy API resource, which is used for specifying
+// policies that should be applied to pods based on the pods' characteristics.
+type PodPolicy struct {
+	unversioned.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
+	ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the pod scheduler policy implemented by this admission controller.
+	// http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status
+	Spec PodPolicySpec `json:"spec,omitempty"`
+}
+
+// PodPolicyList is a list of PodPolicy items.
+type PodPolicyList struct {
+	unversioned.TypeMeta `json:",inline"`
+	// Standard list metadata.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#types-kinds
+	unversioned.ListMeta `json:"metadata,omitempty"`
+
+	// Items is a list of PodPolicy objects.
+	// More info: http://releases.k8s.io/HEAD/docs/design/admission_control_resource_quota.md#admissioncontrol-plugin-resourcequota
+	Items []PodPolicy `json:"items"`
+}
+```
+
+## Implementation plan
+
+1. Create `PodPolicy` API resource
+1. Create admission controller that uses policy defined in `PodPolicy` to set scheduler name annotation
+
+*TODO*: What are the semantics of having multiple `PodPolicy` API resources? "Apply all of them" is not
+ideal since there is an ordering to the rules within one. We could do "apply all of them in
+alphabetical order" or something like that...
+
+*TODO*: Is there a way to specify a singleton `PodPolicy` that would apply across all namespaces?
```

Comment:
Having gone down this route with both pod security policy and authorization policy, we ended up making most of the resources cluster level, and then when we needed to having per namespace overrides.  However, per namespace often wasn't an override - it was an underride - since we want cluster-admins to have primary control.

We debated singleton Quota and Limits and decided it wasn't worth it - we just use "default" and left the door open for others.  It's 100 time saner in REST to model singletons as the only item returned by a list than to start making up rules about which resources are and are not singletons.